### ISSUE TITLE
Conversation log

### ISSUE THREAD

**Issue by ArcEye** [ArcEye/mk-single-module-dir#1]
@zultron 
This is an issue so we can blather on without disturbing lists etc.

**Comment by ArcEye** [comment]
In looking at why the flavour is not detected and defaults to POSIX, I am beginning to wonder about the usefulness of a lot of the code, now that we have stripped down effectively to `rt-preempt` in all modern configurations, with a possible backup of `posix`.

Anyway the error when running rt-preempt kernel is
```
Dec 13 14:03:20 INTEL-i7 rtapi:0: 1:rtapi:28386:rt RTAPI:0 BUG: thread flavors dont match: global 1 rtapi 0
Dec 13 14:03:20 INTEL-i7 rtapi:0: rtapi_app_main(rtapi): -22 Invalid argument
Dec 13 14:03:20 INTEL-i7 rtapi:0: 1:rtapi_app:28386:user init_actions() failed
```

So I thought, how about I run with the $FLAVOR env-var set to `rt-preempt`
`$ export FLAVOR=rt-preempt && machinekit  `
(FAILS JUST THE SAME)

This is a bit strange because the flavor() function checks the env-var first.

Next test is to build with `./configure --with-rt-preempt`
Sure enough this builds and runs perfectly.

So if we tell it a single flavour to build, everything is fine 

Next test is to build with both possible kernels, `/configure --with-rt-preempt --with-posix`
This fails exactly the same way as originally.

**Comment by ArcEye** [comment]
> Thereafter the running flavour can be determined quite simply by grepping uname -r | /proc/version | dmesg and setting accordingly or similar.

In fact I am current playing with a simple script to determine flavour currently and looking to set this at realtime stage.

```
#!/bin/bash

FILE1=$(uname -a)
RT_STR=RT
PREEMPT_STR=PREEMPT
SMP_STR=SMP

if grep -q "$PREEMPT_STR" <<< $FILE1 && grep -q "$RT_STR" <<< $FILE1; then
    KERNEL_FLAVOR=rt-preempt;
else
    if grep -q "$SMP_STR" <<< $FILE1 && (! grep -q "$RT_STR" <<< $FILE1); then
	KERNEL_FLAVOR=posix;
    fi
fi

echo $KERNEL_FLAVOR
```

**Comment by zultron** [comment]
I'm already gleefully beginning to throw away the kernel threads code, partly just to get down to the essentials where I can start updating the user thread build.

The code to determine thread flavor is already there in C, BTW, and should be more reliable than scraping the kernel version string.

The system should build all thread style it's able to, given available headers and libraries, and irrespective of the running kernel. With only one set of modules to build, the extra flavors will come for almost free.

On my phone here, sorry if I missed major points in the above messages. I'll be on this more seriously in a couple of days.

John Morris
Dovetail Automata LLC

________________________________

**Comment by ArcEye** [comment]
Cleaned out some more crud from configure.ac today.

Left some of the Xenomai stuff because it requires careful dissection to remove the hard threads stuff and leave the rest.  Tends to test for xenomai generally first, then divide into userland and rt.

I didn't have time to give it proper attention and don't have a xenomai kernel installed to throw up errors if I make a slip.

**Comment by ArcEye** [comment]
Found the culprit(s) for the posix only running.

One was a missing backslash, causing the use of 2 different shell executions in the Makefile instead of 1.

The other is not quite understood yet, but changing the RTDIR_EXT in Makefile.inc.in from
`$(threads)` to a fixed path of `modules` screwed things up for the build.

Don't do anything with the repo until I have time to put it all into master and test it, hopefully this afternoon.

**Comment by ArcEye** [comment]
The repo now builds and works.
I have squashed down some of the commits to make it tidier.

At present it builds all flavours, because the building of `rtapi_<flavor>_app` and `ulapi_<flavor>` is 
part of the module build.
That is probably the next task, get them built separately, then only build modules once.

The modules overwrite each other and only one set is retained

**Comment by zultron** [comment]
@ArcEye, that's a lot of progress.  I will hack the modules to build only once in the common `modules/` directory.  That shouldn't be too much trouble, and I'll get on it right now.

At least for now, I still believe it's fine to build all flavors detected by `./configure`.  In the short run, reworking the build system is more painful than the few extra seconds of build time, and we can get something working quickly.  In the long run, we can discuss, but one interesting option is to simplify even further by merging the packaging as well.

**Comment by zultron** [comment]
Boy, this build system is seriously fucked.

**Comment by zultron** [comment]
There may be more problems to solve than just folding the modules directories together.  One problem is the `rtapi.so` module needs to be separately built for each flavor, but the non-unique naming will break the flavors at run time.  

The easy way around this is to rename it to `rtapi_$(threads).so`, and teach `rtapi_app_$(threads)` to load it.

The better way around this is to simply fold it back into `rtapi_app_$(threads)`, and do away with the whole `rtapi_switch` ugliness that was never needed as long as there was no unified `rtapi_app`.  I started looking at this option, but didn't get very far.

Also, the `ulapi_$(threads).so` DLLs also need to be merged together, since there's no difference in ULAPI at run time (at least I  think, and hope).

**Comment by bobvanderlinden** [comment]
Good to hear this topic coming up again! I don't use machinekit anymore (not into 3d printing anymore), but at the time I thought cleaning up the build system would benefit the project quite a bit.

My main motivation personally was to make it compatible with non-Debian systems and the motivation for the 3D printer company was to build an image that we could reproduce and upgrade ota reliably. At the time I was looking into buildroot and NixOS.

It would help to get more insight how code is structured as well. I noticed files were being include from all kinds of places (it seemed). CMake did help to get a better understanding for those dependencies.

**Comment by zultron** [comment]
Wow, great to hear support for a build system overhaul!

I've cited many problems with the build system in the past, e.g. why conflating the old build system's `$EMC_HOME` with the newer Autoconf `$prefix` causes problems, how not using Autoconf or CMake with `libtool` necessitates a separate RIP build and prevents running tests as part of package builds, how the Makefiles are clever about building separate RTAPI and ULAPI sources but don't leave room to trivially add new types of build artifacts, and related, how its complexity makes it hard to understand how to make changes while maintaining its original organization.  @bobvanderlinden's point about the shotgun approach to include directories is right on, and also applies to the way `CFLAGS` and `LDFLAGS` are applied in general, and in the current build system, it would be really painful to try to clean that up.

I've made some progress in the C sources to fold the flavors together better than before.  On top of @ArcEye's unification of the modules, I'm almost done unifying the (needlessly) separate ULAPI flavors.  (I've also dropped a couple thousand lines of code that supported kernel threads.)

That leaves only `libexec/rtapi_app_$(threads)` and `rtlib/$(threads)/rtapi.so` that need separate, per-flavor builds.  The latter really needs to be merged into the former.  They were kept separate before simply because `rtapi.ko` *had* to be separate when it was a kernel module, and also e.g. `rtapi_app_rtai` could work with different `rtapi.ko` modules built for different RTAI kernel versions.  Now that we're ditching kernel threads, there's a one-to-one mapping between the two, and they may as well be built into the same executable, and the complicated system with `rtapi_switch` and flavor-specific "hooks" to join them can be mostly eliminated.  (That also leaves an opportunity for later to unify flavors into a single `rtapi_app` that chooses the flavor internally, letting us ditch a lot of the external run-time support apparatus, something that may be very attractive once Xenomai support can be dropped.)

Whether `rtapi_app` and `rtapi.so` are merged or not, unifying these bits should simplify build system requirements quite a lot.  I'm quite interested in revisiting CMake, and I've already looked a bit at @bobvanderlinden's [old branch][1]; thanks for the pointer's to @kinsamanka's [work][2] (I hadn't seen that before).

[1]: https://github.com/bobvanderlinden/machinekit/tree/cmake
[2]: https://github.com/kinsamanka/machinekit/tree/cmake

Semi-OT question, @bobvanderlinden Which non-Debian systems were you using?  I actually started developing LinuxCNC in the first place by wanting to package it for RedHat distros, and my work with RTAPI started with reviving @mbuesch's RT_PREEMPT patches that matched semi-official RT kernel packages.  (LCNC was supposed to be a short-term project, but somehow six years later I'm still developing it and MK!)

**Comment by bobvanderlinden** [comment]
For me personally, NixOS and maybe buildroot. I haven't done much with buildroot since then, but I'm still  a big supporter of NixOS. However, being more inclusive for other distros (not packaging, but just having a sane build process) will potentially result in more contributors. You generally can't expect people to switch their distro just to work on a piece of software.

**Comment by ArcEye** [comment]
> I've made some progress in the C sources to fold the flavors together better than before. On top of @ArcEye's unification of the modules, I'm almost done unifying the (needlessly) separate ULAPI flavors. (I've also dropped a couple thousand lines of code that supported kernel threads.)

Great, you have push access ( or will have when you take up the invitation email from github) so look forward to seeing them.

I started to get reasonably familiar with CMake working on the Machinekit-NG repo.
I didn't agree with the way it was constructed to build, always creating a separate dir with the binaries etc in, because it did not have the includes and Makefiles in it, so was useless for developing.

I made another version, that built as a RIP and was installable from that RIP
https://github.com/machinekit/machinekit-NG/tree/cmake-hal-inrepo

The main question, is whether it is worth the effort?

The single module build and combined rtapi_xx and ulapi_xx, could be done more simply by bending
the existing system to fit.
Depends upon your appetite for complete change.

If we do, there is a strong case for shifting across to a repo based upon machinekit-hal.
Not only is the basis for a CMake system already available, it will be easier to port back into machinekit proper.

The next issue arising from that, is whether now is the time to mothball the pantechnicon machinekit and switch to the machinekit-hal and machinekit-cnc, which already exist and can be combined to become machinekit again.

I am currently keeping both updated, cherry-picking from machinekit commits, but that is not sustainable for ever, nor desirable.

**Comment by ArcEye** [comment]
> If we do, there is a strong case for shifting across to a repo based upon machinekit-hal.
Not only is the basis for a CMake system already available, it will be easier to port back into machinekit proper.

In furtherance to this, I have now cherry-picked all the recent merges (most of them yours @zultron ) into `machinekit-hal` and `machinekit-cnc`.

Once you have committed your changes to this repo, it should be moderately easy to combine everything into an up to date version of `machinekit-hal` with most of CMake already in it, if that is the way we go.

If not it is something I can return to later.
